docs/DeveloperGuide.md

```diff
@@ -2,37 +2,220 @@
+* printBasicWarning(): checks if the user's budget still provides a good cushion (meaning that he/she still has
+more than 25% of his budget left) and prints a warning if not
+* printSeriousWarning(): checks if the user ran out of budget and prints a warning if so
+
+**Step 4**\
+A string containing information about the budget, the net cash from the financial statement and the progress bar
+is built and passed to the method printMultipleText by the Ui.
+
+### String Tokenizer
+
+The String Tokenizer assists in the parsing of user input. It is used to split the user input into a hashmap according to the delimiter/prefix of each command. This allows the developer to more easily customize the commands and add new commands without having to worry about the parsing of the user input.
+
+> **Example**:\
+> `addIncome money /amt 1000 /date 2021-10-10` with prefixes array `{"addIncome", "/amt", "/date"}` will be tokenized into a hash map with the following key-value pairs:\
+> `addIncome: money`\
+> `amt: 1000`\
+> `date: 2021-10-10`
+
+Delimiters can also be specified as optional by adding ":optional" to the end. This allows the user to omit the prefix when entering the command.
+
+#### Implementation
+
+The tokenize function is implemented by the StringTokenizer class. It takes in the user input and the prefixes array and returns a hashmap of the user input.
+
+![](./images/stringTokenizer.png)
+
+**Step 1**\
+Check if all necessary prefixes are present in the user input. If not, throw an exception.
+
+**Step 2**\
+Generate list of PrefixWithPosition (containing the prefix/delimiter and the position in the string) for the delimiters in the user input
+
+**Step 3**\
+Sort the list of PrefixWithPosition by position in order to get the prefixes in the correct order
+
+**Step 4**\
+Create a hashmap of the user input by splitting the string using the prefixes' positions and adding the key-value pairs to the hashmap
+
+### Transaction Filter
+The filter function assists in filtering both expenses and incomes according to the specified criterion from user's input.
+
+#### Implementation
+![](./images/FilterTransaction.png)
+
+**Step 1**\
+The filterBy() method is called, returning a FindParser object
+
+**Step 2**\
+A FilterTransaction object is constructed, with the FindParser object as one of its parameters
+
+**Step 3**\
+From the expenseStatement and incomeStatement, the financialStatement is constructed. 
+
+**Step 4**\
+Execution of the FilterTransaction class calls method findTransaction(), which prints expenses and incomes that match the provided criterion
+
+### File Storage
+
+The FileStorage class in the CashLeh? application facilitates the reading and writing of financial transaction data to and from a file. It simplifies the process of managing user data, making it easily accessible and modifiable through the following methods:
+
+#### Reading Data
+![](./images/readFromFile.png)
```

Comment:
Does "Data read successfully" need to be inside the alt box corresponding to correct transaction format?